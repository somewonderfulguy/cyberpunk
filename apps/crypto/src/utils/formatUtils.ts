import { FiatKey } from '../apiClient/balance/balanceTypes'

const fiatToLocale = new Map<FiatKey, string>([
  ['usd', 'en-US'],
  ['uah', 'uk-UA'],
  ['gbp', 'en-GB'],
  ['eur', 'de-DE'],
  ['pln', 'pl-PL'],
  ['jpy', 'ja-JP']
])

export const formatFiat = (
  value: number,
  currency: FiatKey,
  options?: Intl.NumberFormatOptions
) =>
  new Intl.NumberFormat(fiatToLocale.get(currency), {
    style: 'currency',
    currency,
    ['trailingZeroDisplay' as string]: 'stripIfInteger',
    ...options
  })
    .format(value)
    // if the currency symbol is at the end with space - move to the beginning
    .replace(/(.{1,})(\s€|₴)$/, (match, p1, p2) => p2.trim() + p1.trim())
