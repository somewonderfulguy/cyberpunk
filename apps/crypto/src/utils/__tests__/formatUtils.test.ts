import { formatFiat } from '../formatUtils'

const number = 12345624.78934534
const numberAlt = 12345624

test('formatFiat', () => {
  const result = [
    formatFiat(number, 'usd'),
    formatFiat(numberAlt, 'usd'),
    formatFiat(number, 'eur'),
    formatFiat(numberAlt, 'eur'),
    formatFiat(number, 'uah'),
    formatFiat(numberAlt, 'uah'),
    formatFiat(number, 'gbp'),
    formatFiat(numberAlt, 'gbp'),
    formatFiat(number, 'pln'),
    formatFiat(numberAlt, 'pln'),
    formatFiat(number, 'jpy'),
    formatFiat(numberAlt, 'jpy')
  ]
  expect(result).toMatchSnapshot()
})
