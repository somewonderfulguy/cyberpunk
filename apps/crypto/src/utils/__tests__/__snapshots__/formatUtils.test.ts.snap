// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatFiat 1`] = `
[
  "$12,345,624.79",
  "$12,345,624",
  "€12.345.624,79",
  "€12.345.624",
  "₴12 345 624,79",
  "₴12 345 624",
  "£12,345,624.79",
  "£12,345,624",
  "12 345 624,79 zł",
  "12 345 624 zł",
  "￥12,345,625",
  "￥12,345,624",
]
`;
